keras_preprocessing/image/dataframe_iterator.py

```diff
@@ -102,61 +118,127 @@ def __init__(self,
+        if class_mode not in ["other", "input", None]:
+            df, classes = self._filter_classes(df, y_col, classes)
+            num_classes = len(classes)
+            # build an index of all the unique classes
+            self.class_indices = dict(zip(classes, range(len(classes))))
+        # retrieve only training or validation set
         if self.split:
-            num_files = len(self.df)
+            num_files = len(df)
             start = int(self.split[0] * num_files)
             stop = int(self.split[1] * num_files)
-            self.df = self.df.iloc[start: stop, :]
-        self.filenames = self.df[x_col].tolist()
-
+            df = df.iloc[start: stop, :]
+        # get labels for each observation
         if class_mode not in ["other", "input", None]:
-            classes = self.df[y_col].values
-            self.classes = np.array([self.class_indices[cls] for cls in classes])
-        elif class_mode == "other":
-            self.data = self.df[y_col].values
-            if type(y_col) == str:
+            self.classes = self.get_classes(df, y_col)
+        self.filenames = df[x_col].tolist()
+        # create numpy array of raw input if class_mode="other"
+        if class_mode == "other":
+            self.data = df[y_col].values
+            if isinstance(y_col, str):
                 y_col = [y_col]
-            if "object" in list(self.df[y_col].dtypes):
+            if "object" in list(df[y_col].dtypes):
                 raise TypeError("y_col column/s must be numeric datatypes.")
         self.samples = len(self.filenames)
-        if self.num_classes > 0:
-            print('Found %d images belonging to %d classes.' %
-                  (self.samples, self.num_classes))
+        if class_mode in ["other", "input", None]:
+            print('Found {} images.'.format(self.samples))
         else:
-            print('Found %d images.' % self.samples)
-
+            print('Found {} images belonging to {} classes.'
+                  .format(self.samples, num_classes))
         super(DataFrameIterator, self).__init__(self.samples,
                                                 batch_size,
                                                 shuffle,
                                                 seed)
 
+    def _check_params(self, df, x_col, y_col, classes):
+        # check class mode is one of the currently supported
+        if self.class_mode not in self.allowed_class_modes:
+            raise ValueError('Invalid class_mode: {}; expected one of: {}'
+                             .format(self.class_mode, self.allowed_class_modes))
+        # check that filenames/filepaths column values are all strings
+        if not all(df[x_col].apply(lambda x: isinstance(x, str))):
+            raise ValueError('All values in column x_col={} must be strings.'
+                             .format(x_col))
+        # check labels are string or numeric if class_mode is binary or sparse
```

Comment:
the comment is not right I think. You only check for strings, not numeric.

Reply:
I think it does make sense to support numeric.

Reply:
Since the has_ext option has been removed, It expects file extensions. So doesn't support numeric now.

Reply:
The `x_col` are just the filenames. Sort of obvious check but the feedback to the user with the message is the nice thing here. Since it mentions explicitly the column being used.